compiler/rustc_const_eval/src/interpret/validity.rs

```diff
@@ -1240,6 +1240,14 @@ impl<'rt, 'tcx, M: Machine<'tcx>> ValueVisitor<'tcx, M> for ValidityVisitor<'rt,
                     self.visit_field(val, 0, &self.ecx.project_index(val, 0)?)?;
                 }
             }
+            ty::Pat(_base, pat) => {
+                // When you extend this match, make sure to also add tests to
+                // tests/ui/type/pattern_types/validity.rs
+                match **pat {
+                    // Range patterns are handled fully by looking at the layout
+                    ty::PatternKind::Range { .. } => {},
```

Comment:
Seems worth asserting that `layout.backend_repr` is indeed `Scalar`... skipping the recursion here is a but dubious, we don't do this for anything else.

Reply:
In particular, I think this means that e.g. if a `u32` range pattern type is uninitialized, or a `char` range pattern type is not a valid `char`, we skip the nice error that `try_visit_primitive` would generate, and instead show a more generic error in `visit_scalar`.

Reply:
Hmm  tried recursing, but it doesn't actually do anything better

Reply:
Can you add `char` cases to `pattern_types/validity.rs`, one case where the character is valid but outside the range and one case where it's not even a valid `char`?

Reply:
That did it, yea

Reply:
So that test does show a different error with the extra call to `visit_value`? Great :)

Reply:
Yes